File: OneSignalSDK/onesignal/src/main/java/com/onesignal/GMSLocationController.java
/**
 * Modified MIT License
 * <p>
 * Copyright 2020 OneSignal
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * <p>
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.onesignal;

import android.location.Location;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationListener;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;

class GMSLocationController extends LocationController {

    static final int API_FALLBACK_TIME = 30_000;

    private static GoogleApiClientCompatProxy googleApiClient;
    static LocationUpdateListener locationUpdateListener;

    // Started from this class or PermissionActivity
    static void startGetLocation() {
        initGoogleLocation();
    }

    private static void initGoogleLocation() {
        // Prevents overlapping requests
        if (fallbackFailThread != null)
            return;

        synchronized (syncLock) {
            startFallBackThread();

            if (googleApiClient == null || lastLocation == null) {
                GoogleApiClientListener googleApiClientListener = new GoogleApiClientListener();
                GoogleApiClient googleApiClient = new GoogleApiClient.Builder(classContext)
                        .addApi(LocationServices.API)
                        .addConnectionCallbacks(googleApiClientListener)
                        .addOnConnectionFailedListener(googleApiClientListener)
                        .setHandler(getLocationHandlerThread().mHandler)
                        .build();

                GMSLocationController.googleApiClient = new GoogleApiClientCompatProxy(googleApiClient);
                GMSLocationController.googleApiClient.connect();
            } else if (lastLocation != null)
                fireCompleteForLocation(lastLocation);
        }
    }

    private static int getApiFallbackWait() {
        return API_FALLBACK_TIME;
    }

    private static void startFallBackThread() {
        fallbackFailThread = new Thread(new Runnable() {
            public void run() {
                try {
                    Thread.sleep(getApiFallbackWait());
                    OneSignal.Log(OneSignal.LOG_LEVEL.WARN, "Location permission exists but GoogleApiClient timed out. Maybe related to mismatch google-play aar versions.");
                    LocationController.fireFailedComplete();
                    scheduleUpdate(classContext);
                } catch (InterruptedException e) {
                    // Interruptions expected when connection is made to the api
                }
            }
        }, "OS_GMS_LOCATION_FALLBACK");
        fallbackFailThread.start();
    }

    static void fireFailedComplete() {
        synchronized (syncLock) {
            if (googleApiClient != null)
                googleApiClient.disconnect();
            googleApiClient = null;
        }
    }

    static void onFocusChange() {
        synchronized (syncLock) {
            OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "GMSLocationController onFocusChange!");
            // Google location not initialized or connected yet
            if (googleApiClient == null || !googleApiClient.realInstance().isConnected())
                return;

            if (googleApiClient != null) {
                GoogleApiClient googleApiClient = GMSLocationController.googleApiClient.realInstance();
                if (locationUpdateListener != null)
                    LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, locationUpdateListener);
                locationUpdateListener = new LocationUpdateListener(googleApiClient);
            }
        }
    }

    private static class GoogleApiClientListener implements GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
        @Override
        public void onConnected(Bundle bundle) {
            synchronized (syncLock) {
                PermissionsActivity.answered = false;

                if (googleApiClient == null || googleApiClient.realInstance() == null)
                    return;

                OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "LocationController GoogleApiClientListener onConnected lastLocation: " + lastLocation);
                if (lastLocation == null) {
                    lastLocation = FusedLocationApiWrapper.getLastLocation(googleApiClient.realInstance());
                    OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "LocationController GoogleApiClientListener lastLocation: " + lastLocation);
                    if (lastLocation != null)
                        fireCompleteForLocation(lastLocation);
                }

                locationUpdateListener = new LocationUpdateListener(googleApiClient.realInstance());
            }
        }

        @Override
        public void onConnectionSuspended(int i) {
            fireFailedComplete();
        }

        @Override
        public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
            fireFailedComplete();
        }
    }

    static class LocationUpdateListener implements LocationListener {

        private GoogleApiClient googleApiClient;

        // this initializer method is already synchronized from LocationController with respect to the GoogleApiClient lock
        LocationUpdateListener(GoogleApiClient googleApiClient) {
            this.googleApiClient = googleApiClient;
            init();
        }

        private void init() {
            long updateInterval = BACKGROUND_UPDATE_TIME_MS;
            if (OneSignal.isInForeground())
                updateInterval = FOREGROUND_UPDATE_TIME_MS;

            if (googleApiClient != null) {
                LocationRequest locationRequest = LocationRequest.create()
                        .setFastestInterval(updateInterval)
                        .setInterval(updateInterval)
                        .setMaxWaitTime((long) (updateInterval * 1.5))
                        .setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);

                OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "GMSLocationController GoogleApiClient requestLocationUpdates!");
                FusedLocationApiWrapper.requestLocationUpdates(googleApiClient, locationRequest, this);
            }
        }

        @Override
        public void onLocationChanged(Location location) {
            OneSignal.Log(OneSignal.LOG_LEVEL.DEBUG, "GMSLocationController onLocationChanged: " + location);
            lastLocation = location;
        }
    }

    static class FusedLocationApiWrapper {
        @SuppressWarnings("MissingPermission")
        static void requestLocationUpdates(GoogleApiClient googleApiClient, LocationRequest locationRequest, LocationListener locationListener) {
            try {
                synchronized (GMSLocationController.syncLock) {
                    if (googleApiClient.isConnected())
                        LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, locationListener);
                }
            } catch (Throwable t) {
                OneSignal.Log(OneSignal.LOG_LEVEL.WARN, "FusedLocationApi.requestLocationUpdates failed!", t);
            }
        }

        @SuppressWarnings("MissingPermission")
        static Location getLastLocation(GoogleApiClient googleApiClient) {
            synchronized (GMSLocationController.syncLock) {
                if (googleApiClient.isConnected())
                    return LocationServices.FusedLocationApi.getLastLocation(googleApiClient);
            }
            return null;
        }
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/location/internal/controller/impl/GmsLocationController.kt
package com.onesignal.onesignal.location.internal.controller.impl

import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationListener
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.onesignal.onesignal.core.LogLevel
import com.onesignal.onesignal.core.internal.application.IApplicationLifecycleHandler
import com.onesignal.onesignal.core.internal.application.IApplicationService
import com.onesignal.onesignal.core.internal.common.events.EventProducer
import com.onesignal.onesignal.core.internal.common.suspendifyOnThread
import com.onesignal.onesignal.core.internal.logging.Logging
import com.onesignal.onesignal.location.internal.common.LocationConstants
import com.onesignal.onesignal.location.internal.controller.ILocationController
import com.onesignal.onesignal.location.internal.controller.ILocationUpdatedHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import java.io.Closeable

class GmsLocationController(
    private val _applicationService: IApplicationService
) : ILocationController {
    private val _locationHandlerThread = LocationHandlerThread()
    private val _startStopMutex = Mutex()
    private val _event = EventProducer<ILocationUpdatedHandler>()

    // exist after start and before stop. updates are protected by the startStopMutex
    private var _googleApiClient: GoogleApiClientCompatProxy? = null
    private var _locationUpdateListener: LocationUpdateListener? = null

    // contains the last location received from location services
    private var _lastLocation: Location? = null

    override suspend fun start() : Boolean {
        var self = this
        var wasSuccessful = false

        withContext(Dispatchers.IO) {
            _startStopMutex.withLock {
                if (_googleApiClient != null) {
                    if(_lastLocation != null)
                        _event.fire { it.onLocationChanged(_lastLocation!!) }
                    else {
                        val localLastLocation = getLastLocation()
                        if(localLastLocation != null) {
                            setLocationAndFire(localLastLocation)
                        }
                    }
                    wasSuccessful = true
                } else {
                    try {
                        withTimeout(API_FALLBACK_TIME.toLong()) {
                            val googleApiClientListener = GoogleApiClientListener(self)
                            val googleApiClient =
                                GoogleApiClient.Builder(_applicationService.appContext)
                                    .addApi(LocationServices.API)
                                    .addConnectionCallbacks(googleApiClientListener)
                                    .addOnConnectionFailedListener(googleApiClientListener)
                                    .setHandler(_locationHandlerThread.mHandler)
                                    .build()
                            var proxyGoogleApiClient = GoogleApiClientCompatProxy(googleApiClient)

                            // TODO: google api client has a blocking connect with timeout, use that instead of our withTimeout?
                            var result = proxyGoogleApiClient.blockingConnect()

                            if(result?.isSuccess == true) {
                                if (_lastLocation == null) {
                                    var lastLocation = FusedLocationApiWrapper.getLastLocation(googleApiClient)
                                    if(lastLocation != null) {
                                        setLocationAndFire(lastLocation)
                                    }
                                }

                                // only after the connect do we save
                                self._locationUpdateListener = LocationUpdateListener(_applicationService, self, proxyGoogleApiClient.realInstance)
                                self._googleApiClient = proxyGoogleApiClient
                                wasSuccessful = true
                            }
                            else {
                                Logging.debug("GMSLocationController connection to GoogleApiService failed: (${result?.errorCode}) ${result?.errorMessage}")
                            }
                        }
                    } catch (e: TimeoutCancellationException) {
                        Logging.warn("Location permission exists but GoogleApiClient timed out. Maybe related to mismatch google-play aar versions.")
                    }
                }
            }
        }

        return wasSuccessful
    }

    override suspend fun stop() {
        _startStopMutex.withLock {
            if(_locationUpdateListener != null) {
                _locationUpdateListener!!.close()
                _locationUpdateListener = null
            }

            if(_googleApiClient != null) {
                _googleApiClient!!.disconnect()
                _googleApiClient = null
            }

            _lastLocation = null
        }
    }

    override fun getLastLocation(): Location? {
        val apiInstance = _googleApiClient?.realInstance ?: return null
        return FusedLocationApiWrapper.getLastLocation(apiInstance)
    }

    override fun subscribe(handler: ILocationUpdatedHandler) = _event.subscribe(handler)
    override fun unsubscribe(handler: ILocationUpdatedHandler) = _event.unsubscribe(handler)

    private fun setLocationAndFire(location: Location) {
        Logging.debug("GMSLocationController lastLocation: $_lastLocation")

        _lastLocation = location

        _event.fire { it.onLocationChanged(location) }
    }

    private class GoogleApiClientListener(private val _parent: GmsLocationController) : GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
        override fun onConnected(bundle: Bundle?) {
            Logging.debug("GMSLocationController GoogleApiClientListener onConnected")
        }

        override fun onConnectionSuspended(i: Int) {
            Logging.debug("GMSLocationController GoogleApiClientListener onConnectionSuspended i: $i")
            // the assumption is that GMS will reconnect if it can, so we take no action.
        }

        override fun onConnectionFailed(connectionResult: ConnectionResult) {
            Logging.debug("GMSLocationController GoogleApiClientListener onConnectionSuspended connectionResult: $connectionResult")

            suspendifyOnThread {
                _parent.stop()
            }
        }
    }

    class LocationUpdateListener(
        private val _applicationService: IApplicationService,
        private val _parent: GmsLocationController,
        private val googleApiClient: GoogleApiClient
    ) : LocationListener, IApplicationLifecycleHandler, Closeable {

        private var hasExistingRequest = false

        init {
            if(!googleApiClient.isConnected)
                throw Exception("googleApiClient not connected, cannot listen!")

            _applicationService.addApplicationLifecycleHandler(this)
            refreshRequest()
        }

        override fun onFocus() {
            Logging.log(LogLevel.DEBUG, "LocationUpdateListener.onFocus()")
            refreshRequest()
        }

        override fun onUnfocused() {
            Logging.log(LogLevel.DEBUG, "LocationUpdateListener.onUnfocused()")
            refreshRequest()
        }

        override fun close() {
            _applicationService.removeApplicationLifecycleHandler(this)

            if(hasExistingRequest)
                FusedLocationApiWrapper.cancelLocationUpdates(googleApiClient, this)
        }

        private fun refreshRequest() {
            if(!googleApiClient.isConnected) {
                Logging.warn("Attempt to refresh location request but not currently connected!")
                return
            }

            if(hasExistingRequest) {
                FusedLocationApiWrapper.cancelLocationUpdates(googleApiClient, this)
            }

            val updateInterval = if (_applicationService.isInForeground)
                LocationConstants.FOREGROUND_UPDATE_TIME_MS
            else
                LocationConstants.BACKGROUND_UPDATE_TIME_MS

            val locationRequest = LocationRequest.create()
                .setFastestInterval(updateInterval)
                .setInterval(updateInterval)
                .setMaxWaitTime((updateInterval * 1.5).toLong())
                .setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)
            Logging.debug("GMSLocationController GoogleApiClient requestLocationUpdates!")
            FusedLocationApiWrapper.requestLocationUpdates(googleApiClient, locationRequest, this)
            hasExistingRequest = true
        }

        override fun onLocationChanged(location: Location) {
            Logging.debug("GMSLocationController onLocationChanged: $location")
            _parent.setLocationAndFire(location)
        }
    }

    internal object FusedLocationApiWrapper {
        fun cancelLocationUpdates(googleApiClient: GoogleApiClient, locationListener: LocationListener) {
            LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, locationListener)
        }

        fun requestLocationUpdates(googleApiClient: GoogleApiClient, locationRequest: LocationRequest, locationListener: LocationListener) {
            try {
                if (Looper.myLooper() == null)
                    Looper.prepare()
                if (googleApiClient.isConnected)
                    LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, locationListener)
            } catch (t: Throwable) {
                Logging.warn("FusedLocationApi.requestLocationUpdates failed!", t)
            }
        }

        fun getLastLocation(googleApiClient: GoogleApiClient): Location? {
            if (googleApiClient.isConnected)
                return LocationServices.FusedLocationApi.getLastLocation(googleApiClient)
            return null
        }
    }

    protected class LocationHandlerThread internal constructor() :
        HandlerThread("OSH_LocationHandlerThread") {
        var mHandler: Handler

        init {
            start()
            mHandler = Handler(looper)
        }
    }

    companion object {
        val API_FALLBACK_TIME = 30000
    }
}